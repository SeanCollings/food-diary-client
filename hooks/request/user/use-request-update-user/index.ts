import { useEffect } from 'react';
import { URI_USER_PROFILE } from '@client/constants';
import { userProfileFetcher } from '@client/fetchers';
import { IUserResponse } from '@client/interfaces/user-data.type';
import { useUserContext } from '@store/user-context';
import useSWR from 'swr';
import { useWellnessEntriesContext } from '@store/wellness-entries-context';

export const useRequestUpdateUser = (shouldFetch: boolean) => {
  const { setInitialUser } = useUserContext();
  const { resetWellnessUpdated } = useWellnessEntriesContext();

  const { data, error, isValidating } = useSWR(
    shouldFetch ? [URI_USER_PROFILE] : null,
    userProfileFetcher<IUserResponse>,
    {},
  );

  useEffect(() => {
    if (!isValidating && data?.user) {
      setInitialUser(data.user);
      resetWellnessUpdated();
    }
  }, [data, isValidating, setInitialUser, resetWellnessUpdated]);

  return {
    isLoading: !error && shouldFetch && !data,
    isError: error,
  };
};
