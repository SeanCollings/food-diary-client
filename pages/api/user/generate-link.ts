import { CustomAxiosError } from '@client/interfaces/axios.types';
import { createApiClientSecure } from '@server/api-client';
import { API_SHARE_GENERATE_LINK } from '@server/server.constants';
import { NextApiRequest, NextApiResponse } from 'next';

interface IResponse {
  shareLink?: string;
  ok?: boolean;
  message?: string;
}

const handler = async (
  req: NextApiRequest,
  res: NextApiResponse<IResponse>
) => {
  if (req.method === 'PUT') {
    console.log('GENERATE LINK PUT:');

    try {
      const apiClientSecure = await createApiClientSecure(req);

      const { data } = await apiClientSecure.put(API_SHARE_GENERATE_LINK);

      return res.status(201).json({ ok: true, shareLink: data.shareLink });
    } catch (err) {
      return res.status(500).json({
        ok: false,
        message: (err as CustomAxiosError).response?.data.message,
      });
    }
  }

  return res.status(500).json({ ok: false });
};

export default handler;
