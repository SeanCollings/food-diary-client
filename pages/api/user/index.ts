import { NextApiRequest, NextApiResponse } from 'next';

interface IResponse {
  ok: boolean;
  message?: string;
}

const handler = async (
  req: NextApiRequest,
  res: NextApiResponse<IResponse>
) => {
  if (req.method === 'PATCH') {
    console.log('USER PATCH:', req.body);

    return res.status(201).json({ ok: true });
  }

  return;
};

export default handler;
