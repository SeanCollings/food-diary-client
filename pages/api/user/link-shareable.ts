import axios, { AxiosError } from 'axios';
import { NextApiRequest, NextApiResponse } from 'next';
import { getSession } from 'next-auth/react';

interface IResponse {
  shareLink?: string;
  ok?: boolean;
  message?: string;
}
interface IRequest extends NextApiRequest {
  isShared: boolean;
}

const handler = async (req: IRequest, res: NextApiResponse<IResponse>) => {
  const session = await getSession({ req });

  if (req.method === 'PUT') {
    try {
      const { data } = await axios.put(
        `${process.env.SERVER_HOST}/share/link-shareable`,
        req.body,
        {
          headers: {
            Authorization: `Bearer ${session?.accessToken}`,
          },
        }
      );

      return res.status(201).json({ ok: true, shareLink: data.shareLink });
    } catch (err) {
      return res.status(500).json({
        ok: false,
        message: (
          err as AxiosError<{
            statusCode: number;
            message: string;
            error: string;
          }>
        ).response?.data.message,
      });
    }
  }

  return res.status(500).json({ ok: false });
};

export default handler;
