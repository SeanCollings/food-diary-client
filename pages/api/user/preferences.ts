import { createApiClientSecure } from '@server/api-client';
import { API_USER_PREFERENCES } from '@server/server.constants';
import { AxiosError } from 'axios';
import { NextApiRequest, NextApiResponse } from 'next';

interface IResponse {
  ok: boolean;
  message?: string;
}

const handler = async (
  req: NextApiRequest,
  res: NextApiResponse<IResponse>
) => {
  try {
    const apiClientSecure = await createApiClientSecure(req);

    if (req.method === 'PATCH') {
      console.log('PREFERENCES PATCH:', req.body);

      const { data } = await apiClientSecure.patch(
        API_USER_PREFERENCES,
        req.body
      );

      return res.status(201).json({ ok: true });
    }
  } catch (err) {
    return res.status(500).json({
      ok: false,
      message: (
        err as AxiosError<{
          statusCode: number;
          message: string;
          error: string;
        }>
      ).response?.data.message,
    });
  }

  return res.status(500).json({ ok: false });
};

export default handler;
