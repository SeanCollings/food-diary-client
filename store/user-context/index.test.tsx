import { render, waitFor } from '@testing-library/react';
import { userService } from '@client/services/user.service';
import { FC, useEffect } from 'react';
import {
  IUserModel,
  UserContextProvider,
  initialState,
  useUserContext,
} from '.';

jest.useFakeTimers();
jest.mock('@client/services/user.service');

const TestComponent = () => <div>Test component</div>;

const DisplayUser: FC<{
  user: IUserModel | null;
  userLoggedIn: boolean;
}> = ({ user, userLoggedIn }) => (
  <div>
    <div>User id: {user?.id}</div>
    <div>User name: {user?.name}</div>
    <div>User email: {user?.email}</div>
    <div>User sharelink: {user?.shareLink}</div>
    <div>User preferences: {JSON.stringify(user?.preferences)}</div>
    <div>User stats: {JSON.stringify(user?.stats)}</div>
    <div>User logged in: {userLoggedIn.toString()}</div>
  </div>
);

const MockInitialUserComponent: FC<{ runInitial?: boolean }> = ({
  runInitial,
}) => {
  const { user, userLoggedIn, setInitialUser } = useUserContext();

  useEffect(() => {
    if (runInitial) {
      setInitialUser({
        id: 'initial_id',
        name: 'Initial Name',
        email: 'initial@email.com',
        preferences: {
          showDayStreak: true,
          isProfileShared: false,
          showWeeklyExcercise: true,
        },
        stats: { dayStreak: 5, weeklyExercise: 12, weeklyWater: 2 },
        shareLink: 'mock_sharelink',
      });
    }
  }, [runInitial, setInitialUser]);

  return <DisplayUser user={user} userLoggedIn={userLoggedIn} />;
};

const MockUpdateUser: FC = ({}) => {
  const {
    user,
    userLoggedIn,
    updateUser,
    updatePreferences,
    updateShareLinkPreference,
  } = useUserContext();

  useEffect(() => {
    updateUser({ name: 'New Mock Name' });
    updatePreferences({
      showDayStreak: false,
      showWeeklyExcercise: false,
      showWeeklyWater: false,
    });
    updateShareLinkPreference({ isProfileShared: false });
  }, [updateUser, updatePreferences, updateShareLinkPreference]);

  jest.runAllTimers();

  return <DisplayUser user={user} userLoggedIn={userLoggedIn} />;
};

const MockUpdatShareLinkUser: FC<{ toggleShare?: boolean }> = ({
  toggleShare = true,
}) => {
  const { user, userLoggedIn, updateShareLink } = useUserContext();

  useEffect(() => {
    updateShareLink({ shareLink: 'updated_share_link', toggleShare });
  }, [toggleShare, updateShareLink]);

  return <DisplayUser user={user} userLoggedIn={userLoggedIn} />;
};

describe('store - user-context', () => {
  const mockUserService = jest.mocked(userService);
  const mockError = jest.fn();

  const mockInitialStateUser: IUserModel = {
    id: 'mock_id',
    name: 'Mock User',
    email: 'test@emal.com',
    preferences: {
      isProfileShared: true,
      showDayStreak: true,
      showWeeklyExcercise: true,
      showWeeklyWater: true,
    },
    stats: { dayStreak: 2, weeklyWater: 12 },
  };

  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(mockError);
  });

  beforeEach(() => {
    mockUserService.updateUser.mockResolvedValue({} as any);
    mockUserService.updatePreferences.mockResolvedValue({} as any);
    mockUserService.updateSharePreference.mockResolvedValue({} as any);
    mockUserService.generateLink.mockResolvedValue({} as any);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should return initial state', () => {
    const {
      user,
      userLoggedIn,
      setInitialUser,
      updateUser,
      updatePreferences,
      updateShareLinkPreference,
      updateShareLink,
    } = initialState;

    expect(user).toBeNull();
    expect(userLoggedIn).toBeFalsy();
    expect(setInitialUser({} as any)).toBeNull();
    expect(updateUser({} as any)).toBeNull();
    expect(updatePreferences({} as any)).toBeNull();
    expect(updateShareLinkPreference({} as any)).toBeNull();
    expect(updateShareLink({} as any)).toBeNull();
  });

  it('should render provider with children', () => {
    const { asFragment } = render(
      <UserContextProvider>
        <TestComponent />
      </UserContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with initial-state', () => {
    const { asFragment } = render(
      <UserContextProvider
        initialState={{
          user: mockInitialStateUser,
          userLoggedIn: true,
        }}
      >
        <MockInitialUserComponent />
      </UserContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should set initial user', () => {
    const { asFragment } = render(
      <UserContextProvider>
        <MockInitialUserComponent runInitial />
      </UserContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should update user', async () => {
    const { asFragment } = render(
      <UserContextProvider
        initialState={{
          user: mockInitialStateUser,
        }}
      >
        <MockUpdateUser />
      </UserContextProvider>,
    );

    await waitFor(() => {
      expect(mockUserService.updateUser).toHaveBeenCalled();
      expect(mockUserService.updatePreferences).toHaveBeenCalled();
      expect(mockUserService.updateSharePreference).toHaveBeenCalled();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not update user if user is null', async () => {
    const { asFragment } = render(
      <UserContextProvider
        initialState={{
          user: null,
        }}
      >
        <MockUpdateUser />
      </UserContextProvider>,
    );

    await waitFor(() => {
      expect(mockUserService.updateUser).not.toHaveBeenCalled();
      expect(mockUserService.updatePreferences).not.toHaveBeenCalled();
      expect(mockUserService.updateSharePreference).not.toHaveBeenCalled();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should catch error when updateUser fails', async () => {
    mockUserService.updateUser.mockResolvedValue({
      error: 'mock error occurred',
    });

    render(
      <UserContextProvider
        initialState={{
          user: mockInitialStateUser,
        }}
      >
        <MockUpdateUser />
      </UserContextProvider>,
    );

    await waitFor(() => {
      expect(mockUserService.updatePreferences).toHaveBeenCalled();
      expect(mockUserService.updateSharePreference).toHaveBeenCalled();
      expect(mockUserService.updateUser).toHaveBeenCalled();
    });
    expect(mockError).toHaveBeenCalledWith(
      'Error updating user:',
      'mock error occurred',
    );
  });

  it('should catch error when updatePreferences fails', async () => {
    mockUserService.updatePreferences.mockResolvedValue({
      error: 'mock error occurred',
    });

    render(
      <UserContextProvider
        initialState={{
          user: mockInitialStateUser,
        }}
      >
        <MockUpdateUser />
      </UserContextProvider>,
    );

    await waitFor(() => {
      expect(mockUserService.updatePreferences).toHaveBeenCalled();
      expect(mockUserService.updateSharePreference).toHaveBeenCalled();
      expect(mockUserService.updateUser).toHaveBeenCalled();
    });
    expect(mockError).toHaveBeenCalledWith(
      'Error updating preferences:',
      'mock error occurred',
    );
  });

  it('should catch error when updateSharePreference fails', async () => {
    mockUserService.updateSharePreference.mockResolvedValue({
      error: 'mock error occurred',
    });

    render(
      <UserContextProvider
        initialState={{
          user: mockInitialStateUser,
        }}
      >
        <MockUpdateUser />
      </UserContextProvider>,
    );

    await waitFor(() => {
      expect(mockUserService.updatePreferences).toHaveBeenCalled();
      expect(mockUserService.updateSharePreference).toHaveBeenCalled();
      expect(mockUserService.updateUser).toHaveBeenCalled();
    });
    expect(mockError).toHaveBeenCalledWith(
      'Error updating preferences:',
      'mock error occurred',
    );
  });

  it('should update share-link with toggleShare true and set isProfileShared true', async () => {
    const { asFragment } = render(
      <UserContextProvider
        initialState={{
          user: {
            ...mockInitialStateUser,
            preferences: { isProfileShared: undefined },
          },
        }}
      >
        <MockUpdatShareLinkUser />
      </UserContextProvider>,
    );

    await waitFor(() => {
      expect(mockUserService.updateUser).not.toHaveBeenCalled();
      expect(mockUserService.updatePreferences).not.toHaveBeenCalled();
      expect(mockUserService.updateSharePreference).not.toHaveBeenCalled();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should update share-link if toggleShare is false and not set isProfileShared', async () => {
    const { asFragment } = render(
      <UserContextProvider
        initialState={{
          user: {
            ...mockInitialStateUser,
            preferences: {
              ...mockInitialStateUser.preferences,
              isProfileShared: undefined,
            },
          },
        }}
      >
        <MockUpdatShareLinkUser toggleShare={false} />
      </UserContextProvider>,
    );

    await waitFor(() => {
      expect(mockUserService.updateUser).not.toHaveBeenCalled();
      expect(mockUserService.updatePreferences).not.toHaveBeenCalled();
      expect(mockUserService.updateSharePreference).not.toHaveBeenCalled();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not update sharelink if user is null', async () => {
    const { asFragment } = render(
      <UserContextProvider
        initialState={{
          user: null,
        }}
      >
        <MockUpdatShareLinkUser />
      </UserContextProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
