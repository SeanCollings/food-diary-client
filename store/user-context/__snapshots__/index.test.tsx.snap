// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store - user-context should not update sharelink if user is null 1`] = `
<DocumentFragment>
  <div>
    <div>
      User id: 
    </div>
    <div>
      User name: 
    </div>
    <div>
      User email: 
    </div>
    <div>
      User sharelink: 
    </div>
    <div>
      User preferences: 
    </div>
    <div>
      User stats: 
    </div>
    <div>
      User logged in: false
    </div>
  </div>
</DocumentFragment>
`;

exports[`store - user-context should not update user if user is null 1`] = `
<DocumentFragment>
  <div>
    <div>
      User id: 
    </div>
    <div>
      User name: 
    </div>
    <div>
      User email: 
    </div>
    <div>
      User sharelink: 
    </div>
    <div>
      User preferences: 
    </div>
    <div>
      User stats: 
    </div>
    <div>
      User logged in: false
    </div>
  </div>
</DocumentFragment>
`;

exports[`store - user-context should render provider with children 1`] = `
<DocumentFragment>
  <div>
    Test component
  </div>
</DocumentFragment>
`;

exports[`store - user-context should render with initial-state 1`] = `
<DocumentFragment>
  <div>
    <div>
      User id: mock_id
    </div>
    <div>
      User name: Mock User
    </div>
    <div>
      User email: test@emal.com
    </div>
    <div>
      User sharelink: 
    </div>
    <div>
      User preferences: {"isProfileShared":true,"showDayStreak":true,"showWeeklyExcercise":true,"showWeeklyWater":true}
    </div>
    <div>
      User stats: {"dayStreak":2,"weeklyWater":12}
    </div>
    <div>
      User logged in: true
    </div>
  </div>
</DocumentFragment>
`;

exports[`store - user-context should set initial user 1`] = `
<DocumentFragment>
  <div>
    <div>
      User id: initial_id
    </div>
    <div>
      User name: Initial Name
    </div>
    <div>
      User email: initial@email.com
    </div>
    <div>
      User sharelink: mock_sharelink
    </div>
    <div>
      User preferences: {"showDayStreak":true,"isProfileShared":false,"showWeeklyExcercise":true}
    </div>
    <div>
      User stats: {"dayStreak":5,"weeklyExercise":12,"weeklyWater":2}
    </div>
    <div>
      User logged in: true
    </div>
  </div>
</DocumentFragment>
`;

exports[`store - user-context should update share-link if toggleShare is false and not set isProfileShared 1`] = `
<DocumentFragment>
  <div>
    <div>
      User id: mock_id
    </div>
    <div>
      User name: Mock User
    </div>
    <div>
      User email: test@emal.com
    </div>
    <div>
      User sharelink: updated_share_link
    </div>
    <div>
      User preferences: {"showDayStreak":true,"showWeeklyExcercise":true,"showWeeklyWater":true}
    </div>
    <div>
      User stats: {"dayStreak":2,"weeklyWater":12}
    </div>
    <div>
      User logged in: true
    </div>
  </div>
</DocumentFragment>
`;

exports[`store - user-context should update share-link with toggleShare true and set isProfileShared true 1`] = `
<DocumentFragment>
  <div>
    <div>
      User id: mock_id
    </div>
    <div>
      User name: Mock User
    </div>
    <div>
      User email: test@emal.com
    </div>
    <div>
      User sharelink: updated_share_link
    </div>
    <div>
      User preferences: {"isProfileShared":true}
    </div>
    <div>
      User stats: {"dayStreak":2,"weeklyWater":12}
    </div>
    <div>
      User logged in: true
    </div>
  </div>
</DocumentFragment>
`;

exports[`store - user-context should update user 1`] = `
<DocumentFragment>
  <div>
    <div>
      User id: mock_id
    </div>
    <div>
      User name: New Mock Name
    </div>
    <div>
      User email: test@emal.com
    </div>
    <div>
      User sharelink: 
    </div>
    <div>
      User preferences: {"isProfileShared":false,"showDayStreak":false,"showWeeklyExcercise":false,"showWeeklyWater":false}
    </div>
    <div>
      User stats: {"dayStreak":2,"weeklyWater":12}
    </div>
    <div>
      User logged in: true
    </div>
  </div>
</DocumentFragment>
`;
